rados/read_op_exec.go

```diff
@@ -0,0 +1,114 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <stdlib.h>
+// #include <rados/librados.h>
+//
+import "C"
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+// ReadOpExecStep - step for exec operation code.
+type ReadOpExecStep struct {
+	charPtr *C.char
+	size    C.size_t
+	prval   *C.int
+}
+
+// newExecStepOp - init new *execStepOp.
+func newReadOpExecStep() *ReadOpExecStep {
+	es := &ReadOpExecStep{
+		charPtr: nil,
+		size:    0,
+		prval:   (*C.int)(C.malloc(C.sizeof_int)),
+	}
+	*es.prval = 0
+
+	runtime.SetFinalizer(es, func(es *ReadOpExecStep) {
+		if es != nil {
+			es.free()
+			es.freeBuffer()
+			es = nil
+		}
+	})
+	return es
+}
+
+// free - will be called to free any resources, especially C memory.
+func (es *ReadOpExecStep) free() {
+	if es.prval != nil {
+		C.free(unsafe.Pointer(es.prval))
+		es.prval = nil
+	}
+}
+
+// freeBuffer - releases C allocated buffer. It is separated from es.free() because lifespan of C allocated buffer is
+// longer than lifespan of read operation.
+func (es *ReadOpExecStep) freeBuffer() {
+	if es.charPtr != nil {
+		C.free(unsafe.Pointer(es.charPtr))
+		es.charPtr = nil
+	}
+}
+
+// update - update state operation.
+func (es *ReadOpExecStep) update() error {
+	return getError(*es.prval)
+}
+
+// Bytes - go slice of bytes allocated in C.
+func (es *ReadOpExecStep) Bytes() []byte {
```

Comment:
```suggestion
func (es *ReadOpExecStep) Bytes() ([]byte, error) {
```

I think it would improve the "ergonomics" of this call to return an error in the case that `Bytes` gets called too early, before the `op.Operate` call has been performed.

Reply:
The only way i see is to not call `r.update(readOp, 0)` in `ReadOp.Exec` method -  like it is done [here](https://github.com/ceph/go-ceph/blob/master/rados/read_op_omap_get_vals_by_keys.go#L112). is it ok?

Reply:
If I understand you correctly, yes. The `ReadOp.Operate` call is what calls `operation.update` (see https://github.com/ceph/go-ceph/blob/e0e4115a01dfd0b935abd383fdf1855138811af4/rados/operation.go#L102) which is what calls update for each op step.

If you think about it from a user's workflow point-of-view you can write something like:
```go

rop := CreateReadOp()
defer rop.Release()
exStep, err := rop.Exec(foo, bar, baz)
if err != nil {
   // ... handle error ...
}
foo := exStep.Bytes()  // <-- this is a mistake
err = rop.Operate(ioctx, obj, OperationNoFlag)
if err != nil {
   // ... handle error ...
}
```

The API doesn't prevent you from calling `Bytes` before Operate but that's not valid. So I'd like to see the API be a bit "safer" to use in this case even if it's the *wrong* way to use the API. As you see in the file you linked to:
https://github.com/ceph/go-ceph/blob/e0e4115a01dfd0b935abd383fdf1855138811af4/rados/read_op_omap_get_vals_by_keys.go#L58-L61